### PR TITLE
Refactor httplib: Move debug code to a filter #4440

### PR DESCRIPTION
**Pull request by anoymouscoder**
#4440
https://github.com/beego/beego/compare/develop...anoymouscoder:refactor-filter?expand=1#diff-2d3ddf8be721d25ee0d78c8dc8b3a984800822258a894fdef34d2da749676673R638-R645

@flycash I'm not sure this is the result what you want. Becase I think `logFilter()` is meaningless, it prints unreadable golang object. Maybe this is not the debug output that users expect? Looking forward to your suggestions.

### CODE REVIEW

File: client/httplib/httplib.go [hunk 1]
```diff
@@ -654,3 +634,12 @@ func TimeoutDialer(cTimeout time.Duration, rwTimeout time.Duration) func(net, ad
...
 }
+
+func logFilter(next Filter) Filter {
+	return func(ctx context.Context, req *BeegoHTTPRequest) (*http.Response, error) {
+		log.Println(req)
+		resp, err := next(ctx, req)
+		log.Println(resp)
+		return resp, err
+	}
+}
```
Comment (flycash) [inline client/httplib/httplib.go:645]: Please move this method to a new file.
We don't print `req` and `resp`, and we could use `httputil.DumpRequest` and `httputil.DumpResponse`.

The keypint is that we should check http header content-type, and print the body if we got expected content-type.

For example, if the content-type is application/json, we print body, or we don't print body.

Let's assume that you have a builder and it has two options:
```go
type logFilterBuilder struct{
    PrintableContentType []string // default value are those textual content-type
    MaxBodyLength  int // default 0
}
```
if we the content-type of request(or response) is in `PrintableContentType` array, we print the body.

When we print the body, if `MaxBodyLength <=0` we print whole body, or we print substring of body.

If you are still confused, and prefer online discussion, we could have a short discussion on zoom.

Comment (anoymouscoder) [reply client/httplib/httplib.go:645]: @flycash There are three questiones need your help.
1. What is the relationship between `ShowDebug` and `logFilterBuilder`. As you said, we provide a default filter like Opentracing Filter, user can initialize it by themselves, and add it to `FilterChain`. So `ShowDebug` plays a what kind of  role?
2. User can customize the `PrintableContentType` parameter, do we need to verify whether the parameter really printable? (eg. The default value contains all kinds of printable content types, the user input parameter should be a subset of default.)
3. `logFilter` should print the request processed by `FilterChain` including `doRequest()` and response before filterChain processing?
I am willing to discuss online, when do you have time?

Comment (flycash) [reply client/httplib/httplib.go:645]: 1. I prefer that making `logFilter` as default filter which means that we will register this filter when httplib module initialized. So if `ShowDebug` is true, we log request and response, or `logFilter` do nothing. Another option is that we remove `ShowDebug` field, but user should register this filter by themselves.
2. I think users should be responsible for their choices. We only provide default value of `PrintableContentType`, but don't need to verify it when users change it.
3. `logFilter` should log request before `doRequest` and log response after `doReuest`. In our design, `doRequest` is the last "filter", it looks like: `logFilter.next.next...=doRequest`:
```go
func FilterChain(next) Filter{
    return func() {
     logRequest()
     next() // finally we will run doRequest.
     logResponse()
   }
}
```

I always have time except UTC+8 00:00 - 10:00. Please send me an email to tell me your time.

File: client/httplib/filter/log/filter.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+	"github.com/beego/beego/v2/client/httplib"
+)
+
+// FilterChainBuilder is used to register logFilter to FilterChain
+type FilterChainBuilder struct {
+	PrintableContentTypes  []string               // only print the included request and response content types
+	MaxPrintableBodyLength int                    // print all content when value <= 0
+	Log                    func(v ...interface{}) // custom log function
+	Out                    io.Writer              // destination for output, only take effect when the Log function is nil
```
Comment (flycash) [inline client/httplib/filter/log/filter.go:33]: only keep one filed.
The better way is that you define a NewBuilder function, it looks like
```go
func NewFilterBuilder(opts ...Opt) *FilterChainBuilder {
    res = newDefaultBuilder() // in this method, you will use all default value to build the instance.
    for _, o := range opts {
       o(res)
    }
}
```
Using the `Option` pattern. And then you could make FilterChainBuilder as private(optional).
The default value of `Log` field, it should be `logs.Debug`

File: client/httplib/filter/log/filter.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package log
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
```
Comment (flycash) [inline client/httplib/filter/log/filter.go:21]: do not use this, please use beego/v2/core/logs

**Review by flycash — CHANGES_REQUESTED** [review]
Good!

File: client/httplib/filter/log/filter.go (hunk 3 above)
Comment (anoymouscoder) [reply client/httplib/filter/log/filter.go]: Alouthough I followed your suggestion, but want to know why `log` package can not be used.

File: client/httplib/filter/log/filter.go (hunk 2 above)
Comment (anoymouscoder) [reply client/httplib/filter/log/filter.go]: Considering that other filter builders also use `FilterBuilder`(not private), I think this should be unified.  Using option pattern will indeed be more in line with the Open/Close Principle.

**Comment by anoymouscoder** [comment]
I removed maxPrintableBodyLength because I think maxPrintableContentTyps already controls the printing behavior of the body.  I currently can’t think of a scenario where user only needs to print part of the body.

File: client/httplib/filter/log/filter.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	}
+}
+
+// FilterChain can print the request after FilterChain processing and response before processsing
+func (builder *FilterChainBuilder) FilterChain(next httplib.Filter) httplib.Filter {
+	return func(ctx context.Context, req *httplib.BeegoHTTPRequest) (*http.Response, error) {
+		info := &logInfo{}
+		defer info.print(builder.log)
+		resp, err := next(ctx, req)
+		if err != nil {
```
Comment (flycash) [inline client/httplib/filter/log/filter.go:80]: `info.err = err`  because this error is from user's business, not framework.

**Review by flycash — CHANGES_REQUESTED** [review]

**Review by flycash — APPROVED** [review]

**Pull request merged**